contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,794 @@
+   should be assigned to the pod
+4. Kubelet calls `Allocate` on the matching Device Plugins
+5. The user deletes the pod or the pod terminates
+
+When receiving a pod which requests Devices kubelet is in charge of:
+  * deciding which device to assign to the pod's containers 
+    * Note: This will be decided in the future at the scheduler level as
+      part of the Resource Class proposal
+  * Calling the `Allocate` function with the list of devices
+
+The scheduler is still in charge of filtering the nodes which cannot
+satisfy the resource requests.
+
+## Device Plugin
+
+### Introduction
+
+The device plugin is structured in 5 parts:
+1. Registration: The device plugin advertises it's presence to Kubelet
+2. ListDevices: Kubelet calls the device plugin to list it's devices
+3. HealthCheck: The device plugin returns a stream on which it writes when
+   a device's health changes
+4. Allocate: When creating containers, Kubelet calls the device plugin's
+   `Allocate` function so that it can run device specific instructions (gpu
+    cleanup, QRNG initialization, ...) and instruct Kubelet how to make the
+    device available in the container.
+5. Heartbeat: The device plugin polls every 5s Kubelet to know if it's still
+   alive and if it has to re-issue a Register request (e.g: Kubelet crashed
+   between two heartbeats)
+
+### Registration
+
+When starting the device plugin is expected to make a (client) gRPC call
+to the `Register` function that Kubelet exposes.
+
+The communication between Kubelet is expected to happen only through Unix
+sockets and follow this simple pattern:
+1. The device plugins starts it's gRPC server
+2. The device plugins sends a `RegisterRequest` to Kubelet (through a
+   gRPC request)
+4. Kubelet starts it's Discovery phase and calls `ListDevices` and `HealthCheck`
+5. Kubelet answers to the `RegisterRequest` with a `RegisterResponse`
+   containing any error Kubelet might have encountered
+
+### Unix Socket
+
+Device Plugins are expected to communicate with Kubelet through gRPC
+on an Unix socket.
+When starting the gRPC server, they are expected to create a unix socket
+at the following host path: `/var/run/kubernetes`.
+
+For non bare metal device plugin this means they will have to mount the folder
+as a volume in their pod spec ([see Installation](##installation)).
+
+Device plugins can expect to find the socket to register themselves on
+the host at the following path:
+`/var/run/Kubernetes/kubelet.sock`.
+
+### Protocol Overview
+
+When first registering themselves against Kubelet, the device plugin
+will send:
+  * The name of their unix socket
+  * [The API version against which they were built](#versioning).
+  * Their `Vendor` ID or name of the device plugin
+
+Kubelet answers with the minimum version it supports and whether or
+not there was an error. The errors may include (but not limited to):
+  * API version not supported
+  * A device plugin was already registered for this vendor
```

Comment:
Vendor here can mean a vendor identifier or a device identifier or a combination of both. Can you be a bit more specific?
Why can't there be multiple plugins with the same vendor identifier as long as the devices that they expose are all unique?